pandas/core/arrays/array_.py

```diff
@@ -0,0 +1,129 @@
+import numpy as np
+
+from pandas._libs import lib, tslibs
+
+from pandas.core.dtypes.common import is_extension_array_dtype
+from pandas.core.dtypes.dtypes import registry
+from pandas.core.dtypes.generic import ABCIndexClass, ABCSeries
+
+from pandas import compat
+
+
+def array(data, dtype=None, copy=False):
+    """
+    Create an array.
+
+    .. versionadded:: 0.24.0
+
+    Parameters
+    ----------
+    data : Sequence[object]
+        A sequence of objects. The scalars inside `data` should
+        be instances of the scalar type for `dtype`.
+
+        When `data` is an Index or Series, the underlying array
+        will be extracted from `data`.
+
+    dtype : Union[str, np.dtype, ExtensionDtype], optional
+        The dtype to use for the array. This may be a NumPy
+        dtype or an extension type registered with pandas using
+        :meth:`pandas.api.extensions.register_extension_dtype`.
+
+        If not specified, there are two possibilities:
+
+        1. When `data` is a :class:`Series`, :class:`Index`, or
+           :class:`ExtensionArray`, the `dtype` will be taken
+           from the data.
+        2. Otherwise, pandas will attempt to infer the `dtype`
+           from the data.
```

Comment:
I find this statement a bit misleading, as we actually don't infer from scalars (at least not in the sense of eg how Series does it), we only use numpy's inference? 

If we say that we infer, I would expect those to do the same:

```
In [3]: pd.array([pd.Timestamp("2012-01-01")])                                                                                                                                                                      
Out[3]: array([Timestamp('2012-01-01 00:00:00')], dtype=object)

In [5]: pd.array([pd.Timestamp("2012-01-01")], dtype='datetime64[ns]')                                                                                                                                              
Out[5]: array(['2012-01-01T00:00:00.000000000'], dtype='datetime64[ns]')
```

Reply:
Ah, looking now further down in the implementation :-) I see you do handle this case for Period, so here it indeed infers:

```
In [6]: pd.array([pd.Period('2012-01-01', freq='D')])                                                                                                                                                               
Out[6]: 
<PeriodArray>
['2012-01-01']
Length: 1, dtype: period[D]
```

Reply:
I haven't stated it explicitly, but it would be nice if 3rd parties could eventually hook into this as well. Right now I think it's just Period (and maybe interval?) that get inferred. Maybe timestamps with timezones once DatetimeArray is done.

Reply:
And not Timestamps without timezones?

Reply:
It seems that `lib.infer_dtype` doesn't distinguish the two

```python
In [5]: lib.infer_dtype([pd.Timestamp('2017', tz='utc')])
Out[5]: 'datetime'

In [6]: lib.infer_dtype([pd.Timestamp('2017', tz='US/Central')])
Out[6]: 'datetime'
```

I've added interval, and tests.

Reply:
> Perhaps we should be explicit in the docs for that? 

Yes, I would do that. 
Would we, in the longer term, the inference happening here be the same as the inference happening in `Series` ? Or would that actually add to much corner cases from there we don't want to carry over?

> Though that kinda closes the door to inference for 3rd party arrays.

How would you envision third party arrays participate in inference? That seems a bit difficult in any case (trying out all registered ones, ..?), and IMO more error prone for users (if you forget to import the 3rd party library, you silently get different results)